app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -582,8 +668,39 @@ public void reportDownloadProgress(int totalRead) {
         publishProgress(DataPullTask.PROGRESS_DOWNLOADING, totalRead);
     }
 
+    /**
+     * In order to achieve the impression of the progress bar updating smoothly during the wait
+     * period after a retry response is received, use a timer task to report incremental progress
+     * from the state we were last in to the progress value that the server just reported back to
+     * us, splitting up the wait time into uniform time periods for reporting each unit of progress.
+     */
+    private void startReportingServerProgress() {

```

Comment:
You should be able to lift this into a helper class that has `serverProgressCompletedSoFar` and `lastReportedServerProgressValue` as fields and takes in an `AsyncTask` it can publish results to. You could then also move the retry response parsing code into it. I'm suggesting this idea mainly to reduce the size of `DataPullTask`, which is already large, and isolate the retry code so it is easier to reason about.